File: packages/runtime-core/src/solve/resolve-world.ts
import { copyInto, identity, MAT2X3_STRIDE, multiplyInto } from '../math/affine';
import type { Mat2x3 } from '../math/affine';
import type { Pose } from '../skeleton/pose';
import { TRANSFORM_MODE_NORMAL, worldFromParentByMode } from '../skeleton/transform-mode';

// On-demand world resolution (ADR-0003 section 2). At solve-order step 3, a constraint needs the
// would-be world matrix of a bone (its target, itself, its parent) while the authoritative forward
// pass is still step 4. resolveWorld composes the bone's ancestor chain's CURRENT local transforms
// root-to-bone using the SAME multiply routine as step 4, so the world it produces equals what step 4
// will produce for that bone (modulo float). It reflects every animation value and every local delta
// written by an earlier-running constraint in this frame, and it is a pure function of current local
// state: two calls with no intervening local write return identical results.

// The deepest ancestor chain resolveWorld will walk. Far beyond any real rig depth (rigs are a few
// dozen bones deep at most); sized generously so the walk never reallocates.
const MAX_CHAIN_DEPTH = 256;

// Solver-owned scratch, reused across calls so resolveWorld allocates nothing. The solve is
// single-threaded and resolveWorld is never called re-entrantly (no resolveWorld nests inside another
// resolveWorld), so a module-level stack/accumulator is safe.
const chainStack = new Int32Array(MAX_CHAIN_DEPTH);
const accumulator = new Float64Array(MAT2X3_STRIDE);
const product = new Float64Array(MAT2X3_STRIDE);
const matScratch = new Float64Array(MAT2X3_STRIDE);

// Write bone boneIndex's world matrix into out[outOffset .. outOffset+5]. Allocation-free.
export function resolveWorld(
  pose: Pose,
  boneIndex: number,
  out: Float64Array,
  outOffset: number,
): void {
  const { parentIndices, transformModes, local } = pose;

  // Walk root-ward, pushing indices: chainStack[0] = boneIndex, chainStack[depth-1] = root ancestor.
  let depth = 0;
  let cursor = boneIndex;
  while (cursor >= 0) {
    chainStack[depth] = cursor;
    depth += 1;
    cursor = parentIndices[cursor]!;
  }

  // Accumulate root-to-bone: start at the root ancestor's local, then inherit each descendant's local
  // under its transformMode (the SAME mode dispatch as step 4, so this on-demand world equals the
  // forward-pass world for that bone). multiplyInto / worldFromParentByMode forbid aliasing their output
  // with an input, so we ping-pong accumulator -> product -> accumulator.
  copyInto(accumulator, 0, local, chainStack[depth - 1]! * MAT2X3_STRIDE);
  for (let k = depth - 2; k >= 0; k -= 1) {
    const childIndex = chainStack[k]!;
    const childOffset = childIndex * MAT2X3_STRIDE;
    if (transformModes[childIndex] === TRANSFORM_MODE_NORMAL) {
      multiplyInto(product, 0, accumulator, 0, local, childOffset);
    } else {
      worldFromParentByMode(
        product,
        0,
        accumulator,
        0,
        local,
        childOffset,
        transformModes[childIndex]!,
      );
    }
    copyInto(accumulator, 0, product, 0);
  }

  copyInto(out, outOffset, accumulator, 0);
}

// Convenience wrapper returning the world matrix as a tuple, for callers/tests that want a Mat2x3.
// Allocates one tuple; the zero-allocation path is resolveWorld(out, offset).
export function resolveWorldMat(pose: Pose, boneIndex: number): Mat2x3 {
  resolveWorld(pose, boneIndex, matScratch, 0);
  return [
    matScratch[0]!,
    matScratch[1]!,
    matScratch[2]!,
    matScratch[3]!,
    matScratch[4]!,
    matScratch[5]!,
  ];
}

// The world matrix of a bone's PARENT, or identity for a root. Constraints express their result as a
// local transform relative to this frame.
export function parentWorldMat(pose: Pose, boneIndex: number): Mat2x3 {
  const parent = pose.parentIndices[boneIndex]!;
  return parent < 0 ? identity() : resolveWorldMat(pose, parent);
}

// Read a bone's current local matrix as a tuple.
export function localMat(pose: Pose, boneIndex: number): Mat2x3 {
  const offset = boneIndex * MAT2X3_STRIDE;
  const { local } = pose;
  return [
    local[offset]!,
    local[offset + 1]!,
    local[offset + 2]!,
    local[offset + 3]!,
    local[offset + 4]!,
    local[offset + 5]!,
  ];
}

// Overwrite a bone's local matrix from a tuple (the channel constraints write a full matrix, not
// recomposed channels, so step 4 reproduces the intended world exactly).
export function writeLocalMat(pose: Pose, boneIndex: number, m: Mat2x3): void {
  const offset = boneIndex * MAT2X3_STRIDE;
  const { local } = pose;
  local[offset] = m[0];
  local[offset + 1] = m[1];
  local[offset + 2] = m[2];
  local[offset + 3] = m[3];
  local[offset + 4] = m[4];
  local[offset + 5] = m[5];
}

File: runtimes/unity/Marionette.Runtime.Core/Solve/ResolveWorld.cs
using Marionette.Runtime.Core.MathCore;
using Marionette.Runtime.Core.Skeleton;

namespace Marionette.Runtime.Core.Solve
{
    // On demand world resolution (mirrors packages/runtime-core/src/solve/resolve-world.ts). At solve
    // step 3 a constraint needs the would be world matrix of a bone while the authoritative forward pass
    // is still step 4. ResolveWorld composes the bone's ancestor chain root to bone using the SAME
    // multiply routine as step 4, so the world it produces equals what step 4 will produce.
    internal static class ResolveWorld
    {
        // The deepest ancestor chain the walk will follow. Far beyond any real rig depth.
        private const int MaxChainDepth = 256;

        // Solver owned scratch, reused across calls so ResolveWorld allocates nothing. The solve is single
        // threaded and ResolveWorld is never called re entrantly, so static scratch is safe (the harness
        // disables test parallelization to preserve that contract). Mirrors the module level scratch in
        // resolve-world.ts.
        [System.ThreadStatic]
        private static int[]? _chainStack;

        [System.ThreadStatic]
        private static double[]? _accumulator;

        [System.ThreadStatic]
        private static double[]? _product;

        [System.ThreadStatic]
        private static double[]? _matScratch;

        private static int[] ChainStack => _chainStack ??= new int[MaxChainDepth];

        private static double[] Accumulator => _accumulator ??= new double[Affine.Mat2x3Stride];

        private static double[] Product => _product ??= new double[Affine.Mat2x3Stride];

        private static double[] MatScratch => _matScratch ??= new double[Affine.Mat2x3Stride];

        // Write bone boneIndex's world matrix into output[outOffset .. outOffset + 5]. Allocation free.
        public static void Resolve(Pose pose, int boneIndex, double[] output, int outOffset)
        {
            int[] parentIndices = pose.ParentIndices;
            sbyte[] transformModes = pose.TransformModes;
            double[] local = pose.Local;
            int[] chainStack = ChainStack;
            double[] accumulator = Accumulator;
            double[] product = Product;

            int depth = 0;
            int cursor = boneIndex;
            while (cursor >= 0)
            {
                chainStack[depth] = cursor;
                depth += 1;
                cursor = parentIndices[cursor];
            }

            Affine.CopyInto(accumulator, 0, local, chainStack[depth - 1] * Affine.Mat2x3Stride);
            for (int k = depth - 2; k >= 0; k -= 1)
            {
                int childIndex = chainStack[k];
                int childOffset = childIndex * Affine.Mat2x3Stride;
                if (transformModes[childIndex] == Skeleton.TransformModes.Normal)
                {
                    Affine.MultiplyInto(product, 0, accumulator, 0, local, childOffset);
                }
                else
                {
                    Skeleton.TransformModes.WorldFromParentByMode(
                        product,
                        0,
                        accumulator,
                        0,
                        local,
                        childOffset,
                        transformModes[childIndex]);
                }

                Affine.CopyInto(accumulator, 0, product, 0);
            }

            Affine.CopyInto(output, outOffset, accumulator, 0);
        }

        // The world matrix of the bone as a value type, for the tuple style call sites.
        public static Mat2x3 ResolveMat(Pose pose, int boneIndex)
        {
            double[] scratch = MatScratch;
            Resolve(pose, boneIndex, scratch, 0);
            return Affine.Read(scratch, 0);
        }

        // The world matrix of a bone's PARENT, or identity for a root.
        public static Mat2x3 ParentWorldMat(Pose pose, int boneIndex)
        {
            int parent = pose.ParentIndices[boneIndex];
            return parent < 0 ? Affine.Identity() : ResolveMat(pose, parent);
        }

        public static Mat2x3 LocalMat(Pose pose, int boneIndex) =>
            Affine.Read(pose.Local, boneIndex * Affine.Mat2x3Stride);

        public static void WriteLocalMat(Pose pose, int boneIndex, in Mat2x3 m) =>
            Affine.Write(pose.Local, boneIndex * Affine.Mat2x3Stride, in m);
    }
}
